tests/code_size/hello_world_wasm2js.json

```diff
@@ -1,10 +1,10 @@
 {
   "a.html": 697,
   "a.html.gz": 437,
-  "a.js": 865,
-  "a.js.gz": 503,
+  "a.js": 1131,
```

Comment:
some of these sizes seem larger, do you know why?

Reply:
I think its only the wasm2js tests that show increases.  I guess the cost of creating the "fake" memory in wasm2js is at least as big or little bigger than the savings we get from not creating the memory in the normal JS.

Perhaps there is some way to combine the fake WebAssembly memory in `src/wasm2js.js` with the fake one that wasm2js in binaryen generates to save some bytes? 

I think we can take a look at shrinking the overall wasm2js size as a followup if we think its worth it.   (Or we can default to IMPORTED_MEMORY for wasm2js?

Reply:
It is interesting that this particular wasm2js test regressed more than any other from some reason.   The other wasm2js regressions are only a around ~10 bytes.

Reply:
It looks like this is probably with investigating.     Somehow in this one case some of the JS optimizations are being defeated.

I've attached the a.js file before and after this change.   Maybe you can see what it is?  I suspect its AJSDCE

Reply:
Thanks, reading the outputs, looks like two issues here.

One may indeed be in JSDCE. It should remove `new Int8Array(m);` (without an assignment to anywhere - technically it has side effects, but only of potentially throwing on an invalid input or out of memory). This code should be doing it, so some debug prints might help there: https://github.com/emscripten-core/emscripten/blob/574bce3a9269720d41fc6d855b6b0769c3089225/tools/acorn-optimizer.js#L250-L263

The second issue is that we now have
```javascript
  var b = new Int8Array(a);
  var c = new Int16Array(a);
  var d = new Int32Array(a);
  var e = new Uint8Array(a);
  var f = new Uint16Array(a);
  var g = new Uint32Array(a);
  var h = new Float32Array(a);
  var i = new Float64Array(a);
```
**inside** the wasm2js output. That can't be removed by JSDCE, which doesn't look all the way in there. Maybe wasm2js itself can avoid emitting unnecessary ones, although that would require some tracking of things, and might not be easy. It's possible that fixing the previous issue would be enough to avoid a regression here, so this might not be worth it (but could be a TODO).

Reply:
When I investigated a little last night I think the problem with the outer `new Int8Array(m);` calls is that they are they are part of a function now (updateGlobalBufferAndViews)... this function exist when `JSDCE` is run and then gets inlines by closure.  I suspect that `JSDCE` is perhaps treating these statements differently when they are inside functions?

Should we be running `JSDCE` after closure for this reason?  Or update `JSDCE` to work just as well pre-closure?

Reply:
I think improving JSDCE is probably possible. I can look into that if you want.

I'd prefer not to run JSDCE after closure - that would increase build times and build complexity.

Reply:
I will try to avoid the use of the function in favor a duplicating the code in order to avoid this regression for now.

If you could look into the improving JSDCE and I can re-land the use of the function.

Reply:
So actually it looks like this is really the thing that makes all the difference here:

```
+            var a = new ArrayBuffer(16777216);
+            var b = new Int8Array(a);
+            var c = new Int16Array(a);
+            var d = new Int32Array(a);
+            var e = new Uint8Array(a);
+            var f = new Uint16Array(a);
+            var g = new Uint32Array(a);
+            var h = new Float32Array(a);
+            var i = new Float64Array(a);
```

For some reason I can no longer reproduce the stray `new Int8Array(m);` at the top level.. so I think this change is good to land and we can look at cleaning up the wasm2js boilerplat at the top of the asmFunc somehow has a followup.

Reply:
I'm not opposed to that, but assuming we have a plan for that followup. Just guessing at it, I worry it might not be practical to thread the dependencies everywhere in wasm2js, but hopefully it's actually easier when looking at it.

Reply:
From @juj's response here (https://github.com/emscripten-core/emscripten/pull/12806#discussion_r525538445)  it sounds like this is something we are already suffering from can clusure should really be able to clean up this stuff for us.

Any idea why its not?

Should I default to IMPORTED_MEMORY in wasm2js mode until this regression is fixed?

Reply:
I added IMPORTED_MEMORY to the list of exclusions from this feature for now.

Reply:
I mean WASM2JS